### PR TITLE
Add new decode and encode funtions to support 3+ cards copy

### PR DESCRIPTION
**Pull request by shaobaili3**
Thanks for creating this project and being a Lor lover.

I help @wangyi041228 to open a pull request.

She added a few functions to support 3+ cards copy and also sort the faction lists.

### CODE REVIEW

File: lor_deckcodes/decode.py [hunk 1]
```diff
@@ -27,10 +35,23 @@ def decode_deck(deckcode: str):
...
     all_cards.extend(_decode_card_block(3, data))
+
     # 2 card copies
     all_cards.extend(_decode_card_block(2, data))
+
     # 1 card copies
     all_cards.extend(_decode_card_block(1, data))
+
+    # more cards
+    n = data.tell()
```
Comment (Rafalonso) [inline lor_deckcodes/decode.py:46]: Please use more relevant variable names, the code gets very hard to read with single letter names

File: lor_deckcodes/decode.py [hunk 2]
```diff
@@ -27,10 +35,23 @@ def decode_deck(deckcode: str):
...
+    c = data.read(1)
+    data.seek(n)
+    while c:
+        all_cards.extend(_decode_card_block_(data))
```
Comment (Rafalonso) [inline lor_deckcodes/decode.py:50]: This introduces some code repetition with the called function having the same responsibilities as `_decode_card_block`, can you change the algorithm to use it?

File: lor_deckcodes/encode.py [hunk 3]
```diff
@@ -21,7 +21,7 @@ def _get_set_faction_combinations(cards: List["CardCodeAndCount"]):
 def _encode_card_block(data_stream: BytesIO, cards: List["CardCodeAndCount"]) -> None:
     set_faction_combinations = _get_set_faction_combinations(cards)
     write_varint(data_stream, len(set_faction_combinations))
-
+    set_faction_combinations = sorted(set_faction_combinations, key=lambda y: (len([card for card in cards if card.faction == y[1] and card.set == y[0]]), y[0], y[1]))
```
Comment (Rafalonso) [inline lor_deckcodes/encode.py:24]: This should be part of `_get_set_faction_combinations` and I need more context for this

File: lor_deckcodes/encode.py [hunk 4]
```diff
@@ -31,6 +31,15 @@ def _encode_card_block(data_stream: BytesIO, cards: List["CardCodeAndCount"]) ->
             write_varint(data_stream, faction_card.card_id)
 
 
+def _encode_card_block_(data_stream: BytesIO, cards: List["CardCodeAndCount"]) -> None:
```
Comment (Rafalonso) [inline lor_deckcodes/encode.py:34]: Code repetition

**Review by Rafalonso — CHANGES_REQUESTED** [review]
This changes introduces a lot of code repetition, no tests and ambiguous variable names. Also I'm missing a lot of context for this, why do we need the encode ordering? I have not seen any documentation on this

**Comment by wangyi041228** [comment]
Sorry for my bad codes.

https://github.com/RiotGames/LoRDeckCodes
3.The set/faction lists are ordered by increasing length. The contents of the set/faction lists are ordered alphanumerically.

Source code in C#
https://github.com/RiotGames/LoRDeckCodes/blob/master/LoRDeckCodes/LoRDeckEncoder.cs
line 186

**Comment by wangyi041228** [comment]
_encode_card_block for [1, 2, 3] copies and  [4+] copies are kind of different. They have different sorting rule and out style.

**Review by Rafalonso — APPROVED** [review]
Thank you for your contribution, I have cleaned the code a little bit, I'll release this in a bit

**Pull request merged**